File: MVVMDemo/app/src/main/java/com/example/a14512/mvvmdemo/net/RetrofitManager.java
package com.example.a14512.mvvmdemo.net;

import android.util.Log;

import com.example.a14512.mvvmdemo.mode.News;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

/**
 * @author 14512 on 2018/8/20
 */
public class RetrofitManager {
    private static final String TAG = "RetrofitManager";

    private static final int TIME_OUT = 5;
    private Retrofit mRetrofit;
    private ApiService mApiService;

    private RetrofitManager() {
        mRetrofit = new Retrofit.Builder()
                .client(getOkHttpClient())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .baseUrl(ApiService.BASE_URL)
                .build();
        mApiService = mRetrofit.create(ApiService.class);

    }

    private static class SingleHolder {
        private static final RetrofitManager INSTANCE = new RetrofitManager();
    }

    public static RetrofitManager getInstance() {
        return SingleHolder.INSTANCE;
    }

    private OkHttpClient getOkHttpClient() {
        OkHttpClient okHttpClient = new OkHttpClient();
        OkHttpClient.Builder builder = okHttpClient.newBuilder();
        builder.connectTimeout(TIME_OUT, TimeUnit.SECONDS);
        builder.writeTimeout(TIME_OUT, TimeUnit.SECONDS);
        builder.readTimeout(TIME_OUT, TimeUnit.SECONDS);
        builder.addInterceptor(new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Response originalResponse = chain.proceed(chain.request());
                //这里获取请求返回的cookie
                if (!originalResponse.headers("Set-Cookie").isEmpty()) {
                    if(originalResponse.headers("Set-Cookie").size()==2){
                        CookieUtils.COOKIE = originalResponse.headers("Set-Cookie").get(1);
                        Log.e(TAG, "intercept: 保存cookie"+CookieUtils.COOKIE);
                    }
                }
                return originalResponse;
            }
        });
        builder.addInterceptor(new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request.Builder builder = chain.request().newBuilder();

                if(null!=CookieUtils.COOKIE){
                    builder.addHeader("Cookie", CookieUtils.COOKIE);
                    Log.e(TAG, "intercept: 添加cookie"+ CookieUtils.COOKIE);
                }else{
                    Log.e(TAG, "intercept: 添加空cookie" );
                }

                return chain.proceed(builder.build());
            }
        });

        return okHttpClient;
    }

    private class HttpResultFunc<T> implements Func1<Result<T>, T> {

        @Override
        public T call(Result<T> tHttpResult) {
            if (tHttpResult.getErrorCode().equals("0")) {
                throw new PreDealException(tHttpResult.getErrorMsg());
            }
            return tHttpResult.getData();
        }
    }

    public void getNews(Subscriber<News> subscriber){
        mApiService.getNews("4a0090627cf07a50def18da875165740","10")
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(subscriber);
    }
}

File: WeatherKotlin/app/src/main/java/com/example/a14512/weatherkotlin/network/RetrofitManager.kt
package com.example.a14512.weatherkotlin.network

import com.example.a14512.weatherkotlin.BuildConfig
import com.example.a14512.weatherkotlin.WeatherApplication
import com.example.a14512.weatherkotlin.config.BASE_URL
import com.example.a14512.weatherkotlin.network.cookie.PersistenceCookieJar
import com.example.a14512.weatherkotlin.network.cookie.cache.SetCookieCache
import com.example.a14512.weatherkotlin.network.cookie.persistence.SharePrefsCookiePresistor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit
import kotlin.properties.Delegates

/**
 * @author 14512 on 2018/3/17
 */
class RetrofitManager {

    init {
        apiService = build(BASE_URL).create(ApiService::class.java)
    }

    companion object {
        private var apiService: ApiService by Delegates.notNull()

        fun getInstance(): RetrofitManager {
            return RetrofitManager()
        }

        fun getApiService1(): ApiService {
            if (apiService == null) {
                throw NullPointerException("get apiService must be called after init")
            }
            return apiService
        }

        fun <T> createApi(url: String, tClass: Class<T>): T {
            return build(url).create(tClass)
        }

        private fun build(baseUrl: String): Retrofit {
            var retrofit: Retrofit? = null
            synchronized(Retrofit::class.java) {
                retrofit = Retrofit.Builder()
                        .baseUrl(baseUrl)
                        // 使用OkHttp Client
                        .client(buildOkHttpClient())
                        // 集成RxJava处理
                        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                        // 集成Gson转换器
                        .addConverterFactory(GsonConverterFactory.create())
                        .build()
            }

            return retrofit!!
        }

        /**
         * 构建OkHttpClient
         * @return
         */
        private fun buildOkHttpClient(): OkHttpClient {
            //持久化cookie
            val cookieJar = PersistenceCookieJar(SetCookieCache(),
                    SharePrefsCookiePresistor(WeatherApplication.getContext()))
            val loggingInterceptor = HttpLoggingInterceptor()
            loggingInterceptor.level = if (BuildConfig.DEBUG) HttpLoggingInterceptor.Level.HEADERS
                                        else HttpLoggingInterceptor.Level.NONE
            return OkHttpClient().newBuilder().addInterceptor(loggingInterceptor)
                    .cookieJar(cookieJar)
                    .retryOnConnectionFailure(true)
                    .readTimeout(20, TimeUnit.SECONDS)
                    .writeTimeout(20, TimeUnit.SECONDS)
                    .build()
        }
    }

    /**
     * 网络请求
     */
    fun getOne() {

    }
}